File: Algo/LeadToOffer/26_BinarySearchTreeAndBiDirectList.cpp
class Solution {
public:
    struct TopEndNode {
        struct TreeNode* top;
        struct TreeNode* end;
        TopEndNode(): top(NULL), end(NULL) {}
    };

    TopEndNode* ConvertNode(TreeNode* pRootOfTree) {
        TopEndNode* res = new TopEndNode();
        if (!pRootOfTree)
            return res;

        TopEndNode* list_left = ConvertNode(pRootOfTree->left);
        TopEndNode* list_right = ConvertNode(pRootOfTree->right);
        if (list_left->end) {
            list_left->end->right = pRootOfTree;
            pRootOfTree->left = list_left->end;
        } else
            list_left->top = pRootOfTree;

        if (list_right->top) {
            list_right->top->left = pRootOfTree;
            pRootOfTree->right = list_right->top;
        } else
            list_right->end = pRootOfTree;

        res->top = list_left->top;
        res->end = list_right->end;

        return res;
    }
    TreeNode* Convert(TreeNode* pRootOfTree)
    {
        if(!pRootOfTree)
            return NULL;

        TopEndNode* res = ConvertNode(pRootOfTree);

        return res->top;
    }
};

// dfs
class Solution {
public:
    Node *pre, *head;
    Node* treeToDoublyList(Node* root) {
        if(root==NULL) return NULL;
        dfs(root);
        head->left=pre;
        pre->right=head;
        return head;
    }

    void dfs(Node* cur){
        if(cur==NULL) return;
        dfs(cur->left);
        if(pre!=NULL) pre->right=cur;
        else head=cur;
        cur->left=pre;
        pre=cur;
        dfs(cur->right);
    }
};

File: Algo/LeadToOffer/26_BinarySearchTreeAndBiDirectList.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
"""
class Solution(object):
    def treeToDoublyList(self, root):
        """
        :type root: Node
        :rtype: Node
        """
        if root is None:
            return None

        # To the head of each node
        list_l_head = self.treeToDoublyList(root.left)
        list_r_head = self.treeToDoublyList(root.right)

        # merge left and root
        if list_l_head is None:
            root.left = root
            root.right = root
            list_l_head = root
        else:
            list_l_tail = list_l_head
            while list_l_tail.right != list_l_head:
                list_l_tail = list_l_tail.right
            list_l_tail.right = root
            list_l_head.left = root
            root.right = list_l_head
            root.left = list_l_tail

        # merge left and right
        if list_r_head is not None:
            list_r_tail = list_r_head
            while list_r_tail.right != list_r_head:
                list_r_tail = list_r_tail.right
            list_r_tail.right = list_l_head
            list_l_head.left = list_r_tail
            root.right = list_r_head
            list_r_head.left = root

        return list_l_head